### ISSUE TITLE
Refactor BaseModel class

### ISSUE THREAD

**Issue by davidorme** [ImperialCollegeLondon/virtual_rainforest#363]
### Discussed in https://github.com/ImperialCollegeLondon/virtual_rainforest/discussions/215

<div type='discussions-op-text'>

<sup>Originally posted by **dalonsoa** April 28, 2023</sup>
Don't panic! It is much simpler than it looks :) 

At the moment, to create a new model, you need to derive from the `BaseModel` class, ensure that there are a few attributes present (up to 4 after #211 ), validate that those attributes exist in the child class and that they fulfil certain conditions. It all works and makes sense.

However, it has always looked me as a bit convoluted, in particular the part about requiring the presence of the attributes and the check that they are indeed properties. I'm here to suggest an, I think, cleaner approach exploiting once more the `__init_subclass__` method.

The idea would be to require the presence of certain keyword arguments directly when subclassing, and validate those before assigning them to the class attributes. This would be a simple example:

```python
from abc import ABC

class Base(ABC):
    a: str
    b: int

    @classmethod
    def __init_subclass__(cls, a: str, b: int):
        try:
            cls.a = validate_a(a)
            cls.b = validate_b(b)
        except:
            pass

class Child(Base, a="a", b=25):
    pass

if __name__ == "__main__":
    something = Child()
    print(something.a)
    print(something.b)
```

This enforces the `Child` class to be created with appropriate values for `a` and `b` without having to define any properties in the subclass. You just need to check if the values passed are valid - eg. if they are within certain bounds or something like that - but their presence is guarantee by design of the `__init_subclass__` definition.

In summary, checks like these become unnecessary: https://github.com/ImperialCollegeLondon/virtual_rainforest/blob/5a821952bdb25a044b135532f9f7ad7c161ffebf/virtual_rainforest/core/base_model.py#L304

And if you really, really want them to be `properties`, as you have now, you just make it so in the `BaseModel` class directly, saving code everywhere else:

```python
class Base(ABC):
    _a: str
    _b: int

    @property 
    def a(self):
        return self._a

    @property 
    def b(self):
        return self._b

    @classmethod
    def __init_subclass__(cls, a: str, b: int):
        try:
            cls._a = validate_a(a)
            cls._b = validate_b(b)
        except:
            pass

```

Anyway, just an idea, and certainly not urgent. 
</div>

**Comment by davidorme** [comment]
@dalonsoa - we've started a review process for GH issues and revisited this this week. 

We think this looks good and easy to change - can I just check what the difference between an attribute and a property is here? Is making them a property (as in the last example) purely that we are then using the `setter` (and not `getter`) mechanisms to stop users being able to trivially alter the class attributes of models?

**Comment by dalonsoa** [comment]
Yep. That last part is there just to match your current approach for providing the values, but up to you how you want to proceed. The main suggestion of this issue is to require a few things to be provided when subclassing without the need of checking if things are properties.
